examples/step-62/step-62.cc

```diff
@@ -0,0 +1,1468 @@
+  // The calculation of the mass and stiffness matrices is very expensive. These
+  // matrices are the same for all the frequency steps. The right hand side
+  // vector is also the same for all the frequency steps. We use this class to
+  // store these values and re-use them at each frequency step. The
+  // `PointHistory` class has already been used in step-18.
+
+  template <int dim>
+  class PointHistory
+  {
+  public:
+    PointHistory(unsigned int dofs_per_cell);
+
+  private:
+    unsigned int dofs_per_cell;
+
+  public:
+    // We store the mass and stiffness matrices in the variables
+    // mass_coefficient and stiffness_coefficient. We store as well the
+    // right_hand_side and JxW values which are going to be the same for all the
+    // frequency steps.
+    FullMatrix<std::complex<double>>  mass_coefficient;
+    FullMatrix<std::complex<double>>  stiffness_coefficient;
+    std::vector<std::complex<double>> right_hand_side;
+    std::complex<double>              JxW;
+  };
+
+
+
+  // @sect4{`get_stiffness_tensor` function}
+
+  // This class returns the stiffness tensor of the material. For the sake of
+  // simplicity we consider the stiffness to be isotropic and homogeneous; only
+  // the density $\rho$ depends on the position. As we have previously done in
+  // step-8. The stiffness coefficients $c_{ijkl}$ can be expressed in function
+  // of the two coefficients $\lambda$ and $\mu$. The coefficient tensor reduces
+  // to
+  // @f[
+  //   c_{ijkl}
+  //   =
+  //   \lambda \delta_{ij} \delta_{kl} +
+  //   \mu (\delta_{ik} \delta_{jl} + \delta_{il} \delta_{jk}).
+  // @f]
+  template <int dim>
+  SymmetricTensor<4, dim> get_stiffness_tensor(const double lambda,
+                                               const double mu)
+  {
+    SymmetricTensor<4, dim> stiffness_tensor;
+    for (unsigned int i = 0; i < dim; ++i)
+      for (unsigned int j = 0; j < dim; ++j)
+        for (unsigned int k = 0; k < dim; ++k)
+          for (unsigned int l = 0; l < dim; ++l)
+            stiffness_tensor[i][j][k][l] =
+              (((i == k) && (j == l) ? mu : 0.0) +
+               ((i == l) && (j == k) ? mu : 0.0) +
+               ((i == j) && (k == l) ? lambda : 0.0));
+    return stiffness_tensor;
+  }
+
+
+
+  // @sect3{`ElasticWave` class}
+
+  // Next let's declare the main class of this program. Its structure is very
+  // similar to the step-40 tutorial program. The main differences are:
+  // - The sweep over the frequency vector.
+  // - We save the stiffness and mass matrices in `quadrature_point_history` and
+  // use them for each frequency step.
+  // - We store the measured energy by the probe for each frequency step in the
+  // HDF5 file.
+  template <int dim>
+  class ElasticWave
+  {
+  public:
+    ElasticWave(Parameters<dim> parameters_);
+    ~ElasticWave();
+    void run();
+
+  private:
+    void setup_system();
+    void assemble_system(double omega, bool calculate_quadrature_data);
```

Comment:
here and below, you pass a lot of arguments by value. We have a convention that all input arguments are always marked as `const`. Can you make that change?